File: Dynamic Programming/rectangle-cutting/iterative_dp.cpp
#include <bits/stdc++.h>
using namespace std;

const long long inf = 1e10;
const long long modulo = 1e9 + 7;

int min(int a, int b) {
	if (a < b) return a;
	else return b;
}

int main() {
	int n, m; cin >> n >> m;
	vector<vector<long>> dp(n+1, vector<long>(m+1, 0));

	for(int i = 1; i < n+1; i++) {
		for(int j = 1; j < m+1; j++) {
			if(i == j) {
				dp[i][j] = 0;
			} else {
				// horizontal cut
				long hort = inf;
				for(int x = 1; x < i; x++) {
					hort = min(hort, 1 + dp[x][j] + dp[i-x][j]);
				}

				// vertical cur
				long vert = inf;
				for(int y = 1; y < j; y++) {
					vert = min(vert, 1 + dp[i][y] + dp[i][j-y]);
				}

				dp[i][j] = min(vert, hort);
			}
		}
	}

	cout << dp[n][m] << "\n";
} 

File: Dynamic Programming/rectangle-cutting/iterative_dp.py
import resource, sys
sys.setrecursionlimit(10000)
resource.setrlimit(resource.RLIMIT_STACK, (2**29,-1))

inf = 1e10
modulo = 1e9 + 7

def solution():
    n, m = map(int, input().split())
    dp = [[0] * (m+1) for _ in range(n+1)] 
    for i in range(1, n+1):
        for j in range(1, m+1):
            # vertical cuts
            if i == j:
                dp[i][j] = 0
            else:
                vert = inf
                for y in range(1, j):
                    vert = min(vert, 1 + dp[i][j-y] + dp[i][y])

                # horizontal cuts
                hort = inf
                for x in range(1, i):
                    hort = min(hort, 1 + dp[i-x][j] + dp[x][j])
                dp[i][j] = min(vert, hort)

    return dp[n][m]


print(solution())
